cpp/src/neighbors/detail/cagra/compute_distance.hpp

```diff
@@ -20,303 +20,280 @@
+};
 
-  template <uint32_t DATASET_BLOCK_DIM, uint32_t TEAM_SIZE, cuvs::distance::DistanceType METRIC>
-  __device__ DISTANCE_T compute_similarity(const QUERY_T* const query_ptr,
-                                           const INDEX_T dataset_i,
-                                           const bool valid) const
+template <typename Spec, typename... Specs>
+struct instance_selector<Spec, Specs...> {
+  template <typename DataT, typename IndexT, typename DistanceT, typename DatasetT>
+  static auto select(const cagra::search_params& params,
+                     const DatasetT& dataset,
+                     cuvs::distance::DistanceType metric)
+    -> std::enable_if_t<spec_sound<Spec, DataT, IndexT, DistanceT, DatasetT>,
+                        std::tuple<init_desc_type<DataT, IndexT, DistanceT, DatasetT>, double>>
   {
-    const auto dataset_ptr  = ptr + dataset_i * ld;
-    const unsigned lane_id  = threadIdx.x % TEAM_SIZE;
-    constexpr unsigned vlen = device::get_vlen<LOAD_T, DATA_T>();
-    // #include <raft/util/cuda_dev_essentials.cuh
-    constexpr unsigned reg_nelem = raft::ceildiv<unsigned>(DATASET_BLOCK_DIM, TEAM_SIZE * vlen);
-    raft::TxN_t<DATA_T, vlen> dl_buff[reg_nelem];
-
-    DISTANCE_T norm2 = 0;
-    if (valid) {
-      for (uint32_t elem_offset = 0; elem_offset < dim; elem_offset += DATASET_BLOCK_DIM) {
-#pragma unroll
-        for (uint32_t e = 0; e < reg_nelem; e++) {
-          const uint32_t k = (lane_id + (TEAM_SIZE * e)) * vlen + elem_offset;
-          if (k >= dim) break;
-          dl_buff[e].load(dataset_ptr, k);
-        }
-#pragma unroll
-        for (uint32_t e = 0; e < reg_nelem; e++) {
-          const uint32_t k = (lane_id + (TEAM_SIZE * e)) * vlen + elem_offset;
-          if (k >= dim) break;
-#pragma unroll
-          for (uint32_t v = 0; v < vlen; v++) {
-            const uint32_t kv = k + v;
-            // Note this loop can go above the dataset_dim for padded arrays. This is not a problem
-            // because:
-            // - Above the last element (dataset_dim-1), the query array is filled with zeros.
-            // - The data buffer has to be also padded with zeros.
-            DISTANCE_T d = query_ptr[device::swizzling(kv)];
-            norm2 += dist_op<DISTANCE_T, METRIC>(
-              d, cuvs::spatial::knn::detail::utils::mapping<float>{}(dl_buff[e].val.data[v]));
-          }
-        }
-      }
-    }
-    for (uint32_t offset = TEAM_SIZE / 2; offset > 0; offset >>= 1) {
-      norm2 += __shfl_xor_sync(0xffffffff, norm2, offset);
-    }
-    return norm2;
+    auto s0 = spec_match<Spec, DataT, IndexT, DistanceT, DatasetT>(params, dataset, metric);
+    auto ss = instance_selector<Specs...>::template select<DataT, IndexT, DistanceT, DatasetT>(
+      params, dataset, metric);
+    return std::get<1>(s0) >= std::get<1>(ss) ? s0 : ss;
+  }
+
+  template <typename DataT, typename IndexT, typename DistanceT, typename DatasetT>
+  static auto select(const cagra::search_params& params,
```

Comment:
Yes, you're right. I've added the comments to the struct itself and to other helper functions (but not to `select`, because we have three declarations of it there, which makes it unclear where to add the comment).